brian2/codegen/generators/base.py

```diff
@@ -215,10 +216,27 @@ def has_repeated_indices(self, statements):
                          if index not in ('_idx', '0'))
         return not all_unique
 
+    _translate_cache = {}
     def translate(self, code, dtype):
         '''
         Translates an abstract code block into the target language.
         '''
+        cache_key = _hashable([code, dtype, self.variables,
+                               self.variable_indices,
+                               self.override_conditional_write,
+                               self.allows_scalar_write,
+                               self.iterate_all,
+                               self.template_name,
+                               self.name,
+                               self.owner.name,
+                               {var.owner.name
+                                for var in self.variables.itervalues()
+                                if getattr(var, 'owner', None) is not None},
+                               prefs,
+                               self.__class__])
```

Comment:
This feels unsafe because it's not obvious from this function alone that the outcome of code generation depends solely on these variables, and if you change another function somewhere else to add a new dependency, you might not remember to update this here. As an alternative, what about the following. Make the ``CodeGenerator`` object hashable and give it a hash function. This hash function will then call the hash functions of any objects that are part of the code generator, and like that the code that generates the key is distributed to the place where the items in that key are actually used. In addition, modify the code so that any external variables must be extracted in a single function call which returns a value that is considered part of any key. We can't enforce that in Python, but the big ones would be prefs and ``current_device``, so we could add a switch that makes those temporarily inaccessible, and wrap any code generation functions with something that turns this switch on and then off again at the end. Like that, if we forget and try to access prefs or device in a place it doesn't know about, it will cause an error.